### DISCUSSION TITLE
ZHA performance / network load

### DISCUSSION THREAD

**Discussion by dumpfheimer** [zigpy/zigpy#1296] [ZHA development]
I am continuously trying to increase performance of my Zigbee network.
Lately I have been looking into broadcasts and stumbled across a silabs document that suggested to avoid broadcasts in large networks. I changed most of my scenes back to per-device settings and removed groups wherever possible. I also reduced the number of MTORR's and my network got noticeably more reactive.

I have quite a few Zigbee smart plugs with electrical measurement builtin. I have 2 issues with these devices:

**1. It seems like they are configured to send an update every second, when the power changes:**

[source](https://github.com/home-assistant/core/blob/5bb233998e87ffa0acebdea0d39a6f8a715d614e/homeassistant/components/zha/core/const.py#L303):

```REPORT_CONFIG_RPT_CHANGE = 1```

[source](https://github.com/home-assistant/core/blob/5bb233998e87ffa0acebdea0d39a6f8a715d614e/homeassistant/components/zha/core/const.py#L324-L328):
```
REPORT_CONFIG_OP = (
    REPORT_CONFIG_MIN_INT_OP,
    REPORT_CONFIG_MAX_INT,
    REPORT_CONFIG_RPT_CHANGE,
)
```
[source](https://github.com/home-assistant/core/blob/5bb233998e87ffa0acebdea0d39a6f8a715d614e/homeassistant/components/zha/core/cluster_handlers/homeautomation.py#L67-L77)
```
    REPORT_CONFIG = (
        AttrReportConfig(attr="active_power", config=REPORT_CONFIG_OP),
        AttrReportConfig(attr="active_power_max", config=REPORT_CONFIG_DEFAULT),
        AttrReportConfig(attr="apparent_power", config=REPORT_CONFIG_OP),
        AttrReportConfig(attr="rms_current", config=REPORT_CONFIG_OP),
        AttrReportConfig(attr="rms_current_max", config=REPORT_CONFIG_DEFAULT),
        AttrReportConfig(attr="rms_voltage", config=REPORT_CONFIG_OP),
        AttrReportConfig(attr="rms_voltage_max", config=REPORT_CONFIG_DEFAULT),
        AttrReportConfig(attr="ac_frequency", config=REPORT_CONFIG_OP),
        AttrReportConfig(attr="ac_frequency_max", config=REPORT_CONFIG_DEFAULT),
    )
```

which generates a lot of traffic.

**2. But, the possibly more impactfull issues: It seems like ZHA polls the energy from the device every 30 seconds. This is not really a problem if the device is online and reachable, but if the devices is unplugged, this will result in multiple messages times multiple retries and multiple route discoveries.**

I would need to run more tests to prove this is true, but what I can say is: removing those unplugged devices from ZHA seems to have improved performance of the network.

**This leaves me with the following questions:**

1. Is this somehow in line with anyone who has a deeper understanding?
2. Is it easily possible to only poll from devices that have been seen recently
3. Is it easily possible to change the minimum report time of electrical measurements (or even better, make it user-configurable)?

Edit:
I am using the multi protocol plugin.

**Comment by dmulcahey** [comment]
This is actually on my list to address. The polling of these needs to change. Not only are they polled every 30s they are all polled at the same time. The initial work will be to convert the polling mechanism to something similar to what we do with lights so that each device gets a randomized interval within a range. I may also try to build in some intelligence around the reporting… if we are getting the reports we don’t need to poll.

**Reply by MattWestb** [reply]
Servus @dumpfheimer !
Some user with EZSP and large network is blocking direct children for the coordinator and also not using source routing and its looks working OK with 150+ devices but i dont knowing if they is using Zigbee groups in there network then if broadcast is blocking the routing is being broken in the network and then unicast is not working that the pulling need for working OK

**Reply by dumpfheimer** [reply]
Servus!

Thanks for your input!
@dmulcahey if I've read the logs correctly the requests that poll the data happen after each other, which should keep the load to a minimum.

@MattWestb thanks! The EZSP really is quite capable and my network has been pretty stable for routers since I tried to reduce broadcasts. This might however be a coincidence.

The story is a different one with end devices, though. Especially aqara device seem to drop out of the network lately. I was trying to figure out if source routing might help with this and it turns out it just might. On a side note: setting the source route seems to be defective with EZSP in bellows. The setSourceRoute command does not get called. Which does not really matter because there is not setSourceRoute command implemented in the EZSP schema after version 8 (or so). However there is an undocumented setSourceRoute function in the API, which I added and it seems to work.

I was hoping to find out if aqara devices less frequently drop out of the network when the las hop is from their parent. From what I can se this is not guaranteed without source routing. Also: it seems like the undocumented setSourceRoute only works with source routing disabled (duh, I guess). Does anyone know of an alternative way to force a route for a specific packet? In ZNP this seemed to be really easy.

**Reply by dmulcahey** [reply]
Can you share the details on the undocumented source route functionality? 

The polling actually happens all at once as we use the OOTB HA polling for them. When network sizes scale up this gets very chatty with it happening every 30s. I have some thoughts to fix this part though.

**Reply by dmulcahey** [reply]
Aqara devices are finicky…. Especially the old ones. The Zigbee 3 stuff seems to work well and not drop off the network. Their issues stem from how child device aging works in routing devices. The aqara devices sleep for 55ish minutes and a lot of older routing devices will age them out. This is what causes them to “drop”. https://community.hubitat.com/t/xiaomi-aqara-devices-pairing-keeping-them-connected/623

This link has a bunch of info on this specific issue.

**Reply by dumpfheimer** [reply]
About the setSourceRouting:

This is the code I found:
https://github.com/SiliconLabs/gecko_sdk/blame/gsdk_4.3/protocol/zigbee/app/util/ezsp/command-functions.h#L1718-L1734

I added it to bellows:
https://github.com/dumpfheimer/bellows/blob/dumpfheimer/testing-stuff1/bellows/ezsp/v12/commands.py#L247-L251

And made bellows actually call the function:
https://github.com/dumpfheimer/bellows/blob/c039a29f4c40519702003932146205d0391ad788/bellows/zigbee/application.py#L664-L666

I am guessing that there has not really been a use case for setting the route in bellows yet, because:
The only way to set a route until now has been to use the source route that is derived from messages that are received by bellows, which means that the radio already has that source route in memory. So there would be no benefit in sending it again.

Unless of course the source route has been lost, or one wants to force another source route, which I was trying as I discovered it doesn't work right now.

**Reply by dumpfheimer** [reply]
I forgot to mention that I am using the multi protocol plugin. Before the switch it seemed like the radio was running out of memory and discarding routes which caused delays.

**Reply by dumpfheimer** [reply]
About the Aqara devices:

Please correct me if I'm wrong:
If I understand correctly, Zigbee has defined a way to handle requests to end devices, which is "extended timeouts".
Those messages should remain in a buffer of the parent devices (routers) and be sent to the end device when it checks in.

So, if I was developing a router and needed to decide whether to drop an end device I would:
1. check to see if any messages have been received recently and
2. check if I have any messages in my buffer that should be delivered to that device

Which lead me to believe that source routing end devices back and forcing them over their parent could help keeping them from dropping out.

Also, if the message is in a buffer of a router that is not the parent of the end device, how should that router know when to try to deliver that message? It will never know when the end device has woken up.

I had source routing enabled for some time, but it came with its own issues. It seems like EZSP is very good in finding routes to routers and outperform source routing by far which was why I switched back. This was not a definitive test, but during a couple of hours using source routing I had not a single device dropping off the network, but they were dropping of before and after using 1 to 1 route discovery.

This is why I am eager to find a way to force a route for a packet without enabling source routing.

**Reply by MattWestb** [reply]
I think Aqara / Xaomi is needing one parent that is not timing out them so fast and also being near so its having good signals and not getting problem then pulling its parent.
Then the parent is having getting the pull its only normal network continuation for getting the status to the coordinator but its can also being problems with some routers on the rout that is making it not working well.
So the interesting is with parent the sed is having and if its can keeping it online (i using IKEA routers for my Aqara sensors and its normally working well).
